service/worker/batcher/activities.go

```diff
@@ -0,0 +1,276 @@
+			case err := <-respCh:
+				if err == nil {
+					succCount++
+				} else {
+					errCount++
+				}
+				if succCount+errCount == batchCount {
+					break Loop
+				}
+			case <-ctx.Done():
+				return HeartBeatDetails{}, ctx.Err()
+			}
+		}
+
+		hbd.CurrentPage++
+		hbd.PageToken = resp.NextPageToken
+		hbd.SuccessCount += succCount
+		hbd.ErrorCount += errCount
+		activity.RecordHeartbeat(ctx, hbd)
+
+		if len(hbd.PageToken) == 0 {
+			break
+		}
+	}
+
+	return hbd, nil
+}
+
+func (a *activities) getActivityLogger(ctx context.Context) log.Logger {
+	wfInfo := activity.GetInfo(ctx)
+	return log.With(
+		a.Logger,
+		tag.WorkflowID(wfInfo.WorkflowExecution.ID),
+		tag.WorkflowRunID(wfInfo.WorkflowExecution.RunID),
+		tag.WorkflowNamespace(wfInfo.WorkflowNamespace),
+	)
+}
+
+func (a *activities) getOperationRPS(rps int) int {
+	if rps <= 0 {
+		return a.rps(a.namespace.String())
+	}
+	return rps
+}
+
+func (a *activities) getOperationConcurrency(concurrency int) int {
+	if concurrency <= 0 {
+		return a.concurrency(a.namespace.String())
+	}
+	return concurrency
+}
+
+func startTaskProcessor(
+	ctx context.Context,
+	batchParams BatchParams,
+	taskCh chan taskDetail,
+	respCh chan error,
+	limiter *rate.Limiter,
+	sdkClient sdkclient.Client,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case task := <-taskCh:
+			if isDone(ctx) {
+				return
+			}
+			var err error
+
+			switch batchParams.BatchType {
+			case BatchTypeTerminate:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.TerminateWorkflow(ctx, workflowID, runID, batchParams.Reason)
+					})
+			case BatchTypeCancel:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.CancelWorkflow(ctx, workflowID, runID)
+					})
+			case BatchTypeSignal:
+				err = processTask(ctx, limiter, task,
+					func(workflowID, runID string) error {
+						return sdkClient.SignalWorkflow(ctx, workflowID, runID, batchParams.SignalParams.SignalName, batchParams.SignalParams.Input)
+					})
+			}
+			if err != nil {
+				metricsClient.IncCounter(metrics.BatcherScope, metrics.BatcherProcessorFailures)
+				logger.Error("Failed to process batch operation task", tag.Error(err))
+
+				_, ok := batchParams._nonRetryableErrors[err.Error()]
+				if ok || task.attempts > batchParams.AttemptsOnRetryableError {
+					respCh <- err
+				} else {
+					// put back to the channel if less than attemptsOnError
+					task.attempts++
+					taskCh <- task
+				}
+			} else {
+				metricsClient.IncCounter(metrics.BatcherScope, metrics.BatcherProcessorSuccess)
+				respCh <- nil
+			}
+		}
+	}
+}
+
+func processTask(
+	ctx context.Context,
+	limiter *rate.Limiter,
+	task taskDetail,
+	procFn func(string, string) error,
+) error {
+
+	err := limiter.Wait(ctx)
+	if err != nil {
+		return err
+	}
+	activity.RecordHeartbeat(ctx, task.hbd)
```

Comment:
why do we record a heartbeat here? the main loop does it once already, every time hbd changes. hbd isn't changing across tasks, right?

Reply:
Just want to focus on enable per ns worker for batcher in this change. This is the current logic and I don't want to make changes to it as there is no test coverage. I can update batcher workflow logic in a different PR.